common/common-power.c

```diff
@@ -0,0 +1,165 @@
+/*
+
+Copyright (c) 2017-2019, Feral Interactive
+Copyright (c) 2019, Intel Corporation
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+
+ * Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+ * Redistributions in binary form must reproduce the above copyright
+   notice, this list of conditions and the following disclaimer in the
+   documentation and/or other materials provided with the distribution.
+ * Neither the name of Feral Interactive nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+ */
+
+#define _GNU_SOURCE
+
+#include "common-logging.h"
+#include "common-power.h"
+
+#include <assert.h>
+#include <ctype.h>
+#include <glob.h>
+#include <string.h>
+#include <stdio.h>
+#include <linux/limits.h>
+
+static bool read_file_in_dir(const char *dir, const char *file,
+                             char *dest, size_t n)
+{
+	char path[PATH_MAX];
+	int ret = snprintf(path, sizeof(path), "%s/%s", dir, file);
+	if (ret < 0 || ret >= (int)sizeof(path)) {
+		LOG_ERROR("Path length overrun");
+		return false;
+	}
+
+	FILE *f = fopen(path, "r");
+	if (!f) {
+		LOG_ERROR("Failed to open file for read %s\n", path);
+		return false;
+	}
+
+	size_t read = fread(dest, 1, n, f);
+
+	/* Close before we do any error checking */
+	fclose(f);
+
+	if (read <= 0) {
+		LOG_ERROR("Failed to read contents of %s: (%s)\n",
+		          path, strerror(errno));
+		return false;
+	}
+
+	if (read >= n) {
+		LOG_ERROR("File contained more data than expected %s\n", path);
+		return false;
+	}
+
+	/* Ensure we're null terminated */
+	dest[read] = '\0';
+
+	/* Trim whitespace off the end */
+	while (read > 0 && isspace(dest[read - 1])) {
+		dest[read - 1] = '\0';
+		read--;
+	}
+
+	return true;
+}
+
+static bool get_energy_uj(const char *rapl_name, uint32_t *energy_uj)
+{
+	glob_t glo = { 0 };
+	static const char *path = "/sys/class/powercap/intel-rapl/intel-rapl:0/intel-rapl:0:*";
```

Comment:
I've got no idea how AMD APUs work and I'm pretty sure my overlords would be rather grumpy if I put any serious time into improving AMD APU performance. :wink:  However, if someone does know and wants to wire something similar up, I wouldn't personally be opposed.  I suspect that the sysfs lookups for AMD would be completely different from the ones we do for Intel.  The real question is if AMD measures energy since boot or some sort of instantaneous measurement of power.  For that matter, I don't even know if they have something like RAPL.  I tried to put the really intel-specific bits (the RAPL lookups) in helpers so that they wouldn't get in the way if someone wanted to also add support for some other hardware.

Reply:
That's as expected, no worries Jason. I'll take a look myself when I get a chance, but it'll be no block to this PR.

Reply:
At least on Intel HadesCanyon, AMDGPU kernel module provides Vega H power usage directly as Watts in "amdgpu_pm_info" file under debugfs.